gordo_components/dataset/datasets.py

```diff
@@ -86,42 +72,160 @@ def __init__(
+            resample method is provided then the resulting dataframe will have names
+            identical to the names of the series it got in. If several
+            aggregation-methods are provided then the resulting dataframe will
+            have a multi-level column index, with the series-name as the first level,
+            and the aggregation method as the second level.
+            See :py:func::`pandas.core.resample.Resampler#aggregate` for more
+            information on possible aggregation methods.
+        kwargs: dict
+            kwargs to pass to :class:`RawTimeSeriesDataset`
+        """
+        self.resolution = resolution
+        self.aggregation_methods = aggregation_methods
+        super().__init__(**kwargs)
+
+    def get_data(self) -> Tuple[pd.DataFrame, Optional[pd.DataFrame]]:
+        series_iter: Iterable[pd.Series] = self.data_provider.load_series(
+            from_ts=self.from_ts,
+            to_ts=self.to_ts,
+            tag_list=list(set(self.tag_list + self.target_tag_list)),
+        )
+        data: pd.DataFrame = self.join_timeseries(
+            series_iter,
+            self.from_ts,
+            self.to_ts,
+            self.resolution,
+            aggregation_methods=self.aggregation_methods,
+        )
+        if self.row_filter:
+            data = pandas_filter_rows(data, self.row_filter)
+        return self._X_y_from_dataframe(data)
+
+    def get_metadata(self):
+        metadata = super().get_metadata()
+        metadata.update(
+            {
+                "aggregation-methods": self.aggregation_methods,
+                "resolution": self.resolution,
+            }
+        )
+        return metadata
+
+
+class ShiftedTimeSeriesDataset(TimeSeriesDataset):
+    def __init__(self, shift: int = 1, *args, **kwargs):
+        """
+        Return the same data as given by :class:`TimeSeriesDataset`
+        but offset the ``y`` such that the Nth index of ``y`` relates to the
+        datetime in index N-1 index of ``X``.
+
+        Used if you want to train a model to predict future time step values.
+
+        Parameters
+        ----------
+        shift: int
+            Amount to offset the target column.
+        args: tuple
+            Any arguments to supply which are passed onto :class:`TimeSeriesDataset`
+        kwargs: dict
+            Any keyword arguments to be passed into :class:`TimeSeriesDataset`
+
+        Examples
+        --------
+        If the ``resolution`` is 10 minutes and a provided ``shift`` is 1, then
+        the target value(s) for any sample of ``X`` would have the target value(s)
+        10 minutes into the future.
+        """
+        self.shift = shift
+        super().__init__(*args, **kwargs)
+
+    @property
+    def shift(self) -> int:
+        return self._shift
+
+    @shift.setter
+    def shift(self, value: int):
+        if value < 0:
+            raise ValueError("Shift must be greater than 0")
```

Comment:
Idk, does it in some weird world make sense to not predict, but actually look back instead? So given a state of the sensors try to tell what the previous states were? Should we dissalow that in this dataset? Maybe rather log a warning or something? IDK

Reply:
Let's go for it! :tada: